### PR TITLE
Improve form state handling during loading and when errors occur

### PR DESCRIPTION
**Pull request by rafenden**
This PR includes only API changes.

From now on, redirections, form loading state and error state will now be handled internally, so there is no need for using `LoadingBox` or calling `window.location.href`.

Before:

```
function AddCompanyForm () {
  const [isSubmitting, setIsSubmitting] = useState(false)

  const onSubmit = async (values) => {
    setIsSubmitting(true)
    try {
      const { data } = await axios.post('/create-company', values)
      window.location.href = `/companies/${data.id}`
    } catch (error) {
      // todo handle error
      setIsSubmitting(false)
    }
  }

  return (
    <Form onSubmit={onSubmit}>
      <LoadingBox loading={isSubmitting}>
        <FieldInput label="Name" type="text" />
      </LoadingBox>
    </Form>
  )
}
```

After:

```
function AddCompanyForm () {
  const onSubmit = async (values) => {
    const { data } = await axios.post('/create-company', values)
    return `/companies/${data.id}`
  }

  return (
    <Form onSubmit={onSubmit}>
      {form => (
        <>
          {form.submissionError && (
            <StatusMessage>There was a problem when submitting the form, try again later.</StatusMessage>
          )}

          <FieldInput label="Name" type="text" />
        </>
      )}
    </Form>
  )
}
```

### CODE REVIEW

File: src/forms/elements/__tests__/Form.test.jsx [hunk 1]
```diff
@@ -74,19 +135,24 @@ describe('Form', () => {
...
+          }}
         </Form>,
       )
-      wrapper.simulate('submit')
-      expect(onSubmitSpy).toHaveBeenCalledTimes(0)
+    })
+
+    test('should display the loading indicator', () => {
+      const loader = wrapper.find(LoadingBox)
+      expect(loader.prop('loading')).toBeTruthy()
```
Comment (paulgain) [inline src/forms/elements/__tests__/Form.test.jsx]: `loader.prop('loading')` returns a `boolean` so wouldn't `toBe(true)` be more appropriate compared to `toBeTruthy`

File: src/forms/hooks/__tests__/useForm.test.jsx [hunk 2]
```diff
@@ -668,4 +673,95 @@ describe('useForm', () => {
...
+      const hook = renderHook(() => useForm())
+
+      await act(async () => {
+        await hook.result.current.setIsLoading(true)
+        formState = hook.result.current
+      })
+    })
+
+    test('should set "isSubmitted" to true', () => {
+      expect(formState.isLoading).toBeTruthy()
```
Comment (paulgain) [inline src/forms/hooks/__tests__/useForm.test.jsx]: `toBe(true)` as it's a boolean?

File: src/forms/hooks/__tests__/useForm.test.jsx [hunk 3]
```diff
@@ -668,4 +673,95 @@ describe('useForm', () => {
...
+      })
+    })
+
+    test('should redirect the page', () => {
+      expect(window.location.assign).toHaveBeenCalledTimes(1)
+      expect(window.location.assign).toHaveBeenCalledWith('http://example.com')
+    })
+
+    test('should keep isLoading with truthy value', () => {
+      expect(formState.isLoading).toBeTruthy()
```
Comment (paulgain) [inline src/forms/hooks/__tests__/useForm.test.jsx]: `toBe(true)` as it's a boolean?

File: src/forms/hooks/__tests__/useForm.test.jsx [hunk 4]
```diff
@@ -668,4 +673,95 @@ describe('useForm', () => {
...
+        formState = hook.result.current
+      })
+    })
+
+    test('should set "isLoading" to false', () => {
+      expect(formState.isLoading).toBeFalsy()
+    })
+
+    test('should set "isSubmitted" to true', () => {
+      expect(formState.isSubmitted).toBeTruthy()
```
Comment (paulgain) [inline src/forms/hooks/__tests__/useForm.test.jsx]: `toBe(true)` as it's a boolean?

File: src/forms/hooks/__tests__/useForm.test.jsx [hunk 5]
```diff
@@ -668,4 +673,95 @@ describe('useForm', () => {
...
+        formState = hook.result.current
+      })
+    })
+
+    test('should set "submissionError"', () => {
+      expect(formState.submissionError).toEqual(error)
+    })
+
+    test('should set "isLoading" to false', () => {
+      expect(formState.isLoading).toBeFalsy()
```
Comment (paulgain) [inline src/forms/hooks/__tests__/useForm.test.jsx]: `toBe(false)` as it's a boolean?

File: src/forms/hooks/__tests__/useForm.test.jsx [hunk 6]
```diff
@@ -668,4 +673,95 @@ describe('useForm', () => {
...
+    })
+
+    test('should NOT set "isSubmitted" to true', () => {
+      expect(formState.isSubmitted).toBeFalsy()
```
Comment (paulgain) [inline src/forms/hooks/__tests__/useForm.test.jsx]: `toBe(true)` as it's a boolean?

File: src/forms/hooks/__tests__/useForm.test.jsx [hunk 7]
```diff
@@ -668,4 +673,95 @@ describe('useForm', () => {
...
+      const hook = renderHook(() => useForm({ onSubmit }))
+
+      await act(async () => {
+        await hook.result.current.goForward()
+        formState = hook.result.current
+      })
+    })
+
+    test('should set "isLoading" to false', () => {
+      expect(formState.isLoading).toBeFalsy()
```
Comment (paulgain) [inline src/forms/hooks/__tests__/useForm.test.jsx]: `toBe(false)` as it's a boolean?

File: src/forms/hooks/__tests__/useForm.test.jsx [hunk 8]
```diff
@@ -668,4 +673,95 @@ describe('useForm', () => {
...
+      expect(window.location.assign).toHaveBeenCalledTimes(1)
+      expect(window.location.assign).toHaveBeenCalledWith('http://example.com')
+    })
+
+    test('should keep isLoading with truthy value', () => {
+      expect(formState.isLoading).toBeTruthy()
+    })
+
+    test('should keep isSubmitted with falsy value', () => {
+      expect(formState.isSubmitted).toBeFalsy()
```
Comment (paulgain) [inline src/forms/hooks/__tests__/useForm.test.jsx]: `toBe(false)` as it's a boolean?

File: src/forms/hooks/__tests__/useForm.test.jsx (hunk 6 above)
Comment (rafenden) [reply src/forms/hooks/__tests__/useForm.test.jsx]: Good point, code updated!

File: src/forms/hooks/__tests__/useForm.test.jsx (hunk 5 above)
Comment (rafenden) [reply src/forms/hooks/__tests__/useForm.test.jsx]: Good point, code updated!

File: src/forms/hooks/__tests__/useForm.test.jsx (hunk 3 above)
Comment (rafenden) [reply src/forms/hooks/__tests__/useForm.test.jsx]: Good point, code updated!

File: src/forms/hooks/__tests__/useForm.test.jsx (hunk 8 above)
Comment (rafenden) [reply src/forms/hooks/__tests__/useForm.test.jsx]: Good point, code updated!

File: src/forms/hooks/__tests__/useForm.test.jsx (hunk 7 above)
Comment (rafenden) [reply src/forms/hooks/__tests__/useForm.test.jsx]: Good point, code updated!

File: src/forms/hooks/__tests__/useForm.test.jsx (hunk 4 above)
Comment (rafenden) [reply src/forms/hooks/__tests__/useForm.test.jsx]: Good point, code updated!

**Pull request merged**